go/dashboard/dashboard.py

```
@@ -0,0 +1,156 @@
+    else:
+        items = enumerate(collection)
+    return items
+
+
+def visit_dicts(collection, fn):
+    is_mutable = is_mutable_collection(collection)
+
+    for key, value in collection_items(collection):
+        if is_collection(value):
+            visit_dicts(value, fn)
+
+            if is_mutable and isinstance(value, dict):
+                collection[key] = fn(value)
+
+
+def ensure_handler_fields(*fields):
+    def decorator(fn):
+        @wraps(fn)
+        def wrapper(self, target):
+            missing_fields = [f for f in fields if f not in target]
+            if missing_fields:
+                raise DashboardParseError(
+                    "Dashboard layout handler '%s' is missing fields: %s" %
+                    (fn.__name__, missing_fields))
+            return fn(self, target)
+        return wrapper
+    return decorator
+
+
+class DashboardSyncError(Exception):
+    """
+    Raised when we fail to sync the dashboard with diamondash.
+    """
+
+
+class DashboardParseError(Exception):
+    """
+    Raised when dashboard data cannot be parsed into something that can be
+    given to diamondash.
+    """
+
+
+class Dashboard(object):
+    def __init__(self, name, title, layout):
+        self.diamondash_api = client.get_diamondash_api()
+        self.name = name
+        self.title = title
+        self.layout = layout
+        self.config = None
+
+    def _raw_serialize(self):
+        return {
+            'name': self.name,
+            'title': self.title,
+            'widgets': self.layout.serialize()
+        }
+
+    def sync(self):
+        """
+        Ensures the dashboard exists on diamondash's side
+        """
+        try:
+            raw_config = self._raw_serialize()
+            self.config = self.diamondash_api.replace_dashboard(raw_config)
+        except Exception, e:
+            raise DashboardSyncError("Dashboard sync failed: %s" % e)
+
+    def serialize(self):
+        return self.config
```

Wondering whether this should raise an error if `self.config` is still `None`? And whether maybe it should be named something else? What is returned by `replace_dashboard`?